platforms/allwinner-d1/core/src/ccu.rs

```diff
@@ -0,0 +1,272 @@
+//! This module provides a higher-level interface for the Clock Controller Unit (CCU).
+use d1_pac::CCU;
+use d1_pac::DMAC;
+use d1_pac::{SMHC0, SMHC1, SMHC2};
+use d1_pac::{SPI0, SPI_DBI};
+use d1_pac::{TWI0, TWI1, TWI2, TWI3};
+use d1_pac::{UART0, UART1, UART2, UART3, UART4, UART5};
+
+pub struct Ccu {
+    ccu: CCU,
+}
+
+/// Trait to be implemented for module clocks that can be gated and reset
+pub trait BusGatingResetRegister {
+    /// Enable or disable the clock reset bit
+    fn gating(ccu: &mut CCU, pass: bool);
+    /// Enable or disable the clock gating bit
+    fn reset(ccu: &mut CCU, deassert: bool);
+}
+
+// TODO: should this move into the `Clint`?
+fn sdelay(delay_us: usize) {
+    let clint = unsafe { crate::clint::Clint::summon() };
+    let t = clint.get_mtime();
+    // TODO: confirm clock source of mtime
+    while clint.get_mtime() < (t + 24 * delay_us) {
+        core::hint::spin_loop()
+    }
+}
+
+impl Ccu {
+    #[must_use]
+    pub fn new(ccu: CCU) -> Self {
+        Self { ccu }
+    }
+
+    #[must_use]
+    pub fn release(self) -> CCU {
+        self.ccu
+    }
+
+    /// De-assert the reset bit and enable the clock gating bit for the given module
+    pub fn enable_module<MODULE: BusGatingResetRegister>(&mut self) {
```

Comment:
Misc note, you could enforce that the caller "owns" the peripheral by doing this instead (same for disable), which would also mean we don't have to turbofish the call either:

```suggestion
    pub fn enable_module<MODULE: BusGatingResetRegister>(&mut self, _mod: &mut MODULE) {
```

Reply:
Should we require a mutable reference to be sure of ownership?
It's just a bit more inconvenient, because then we have to change the function parameters to be as `mut` as well, e.g.:

```diff
- pub unsafe fn kernel_uart(ccu: &mut Ccu, gpio: &mut GPIO, uart0: UART0) -> Uart {
+ pub unsafe fn kernel_uart(ccu: &mut Ccu, gpio: &mut GPIO, mut uart0: UART0) -> Uart {
```

Reply:
Up to you! Just providing context if that helps, if it means we need to summon peripherals everywhere, then I agree it might be more trouble than it is worth.

Reply:
I think it's a good idea to pass in the module to enforce ownership, but I don't know if the reference needs to be mutable.
We don't actually modify any registers from the module itself, so from that point of view I think
`_mod: &MODULE` is enough, but maybe that does not really *enforce* the ownership?

Reply:
IF you did not use unsafe anywhere else to create a second instance of a peripheral, then it does enforce ownership. A mutable reference can only be created once, from something owned, and the PAC model will only ever (safely) give you ownership of the peripheral once.

Reply:
Yup, I understand that. I'm going to rephrase, because I don't think it was clear what my *actual* question was.
I think there are two options:
```rust
pub unsafe fn kernel_uart(ccu: &mut Ccu, gpio: &mut GPIO, mut uart0: UART0) -> Uart {
    // Enable UART0 clock.
    ccu.enable_module(&mut uart0);
    ...
}
```
and
```rust
pub unsafe fn kernel_uart(ccu: &mut Ccu, gpio: &mut GPIO, uart0: UART0) -> Uart {
    // Enable UART0 clock.
    ccu.enable_module(&uart0);
    ...
}
```
The downside of option 1 is the additional `mut` in the parameter list, to be able to take a mutable reference.
But then you are sure that the caller has a unique reference to the module.

Then my question was, does option 1 prove/enforce ownership better than option 2?
The caller could also be passing a mutable reference that it received from its caller, so then it also doesn't own the module.
In terms of safety, I don't think there is a difference between both options, since you have the `&mut Ccu`, which is the register block that is actually being modified.